### PR TITLE
IGNITE-22805 Implement REST API for initiating CMG repair

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-22805

### CODE REVIEW

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/recovery/system/ResetClusterRequest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import io.swagger.v3.oas.annotations.media.Schema;
+import java.util.List;
+import java.util.Objects;
+import org.apache.ignite.internal.tostring.S;
+
+/** Request to reset cluster. */
+@Schema(description = "Reset cluster.")
+public class ResetClusterRequest {
+    @Schema(description = "Names of the proposed CMG node names.")
+    private final List<String> cmgNodeNames;
```
Comment (tkalkirill) [inline modules/rest-api/src/main/java/org/apache/ignite/internal/rest/api/recovery/system/ResetClusterRequest.java:32]: Will `toString` print a list?

File: modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/ClusterResetExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+import jakarta.inject.Singleton;
+import org.apache.ignite.internal.disaster.system.exception.ClusterResetException;
+import org.apache.ignite.internal.lang.IgniteInternalException;
+import org.apache.ignite.internal.rest.api.Problem;
+import org.apache.ignite.internal.rest.constants.HttpCode;
+import org.apache.ignite.internal.rest.problem.HttpProblemResponse;
+
+/**
+ * Handles {@link ClusterResetExceptionHandler} and represents it as a rest response.
+ */
```
Comment (tkalkirill) [inline modules/rest-api/src/main/java/org/apache/ignite/internal/rest/exception/handler/ClusterResetExceptionHandler.java:33]:
```suggestion
/** Handles {@link ClusterResetExceptionHandler} and represents it as a rest response. */
```

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryController.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.ignite.internal.disaster.system.SystemDisasterRecoveryManager;
+import org.apache.ignite.internal.rest.ResourceHolder;
+import org.apache.ignite.internal.rest.api.recovery.system.ResetClusterRequest;
+import org.apache.ignite.internal.rest.api.recovery.system.SystemDisasterRecoveryApi;
+import org.apache.ignite.internal.rest.exception.handler.ClusterResetExceptionHandler;
+import org.apache.ignite.internal.rest.exception.handler.IgniteInternalExceptionHandler;
+
+/**
+ * Controller for system groups disaster recovery.
+ */
```
Comment (tkalkirill) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryController.java:32]:
```suggestion
/** Controller for system groups disaster recovery. */
```

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryController.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    @Override
+    public CompletableFuture<Void> reset(ResetClusterRequest command) {
+        return systemDisasterRecoveryManager.resetCluster(command.cmgNodeNames());
+    }
+
+    @Override
+    public void cleanResources() {
+        systemDisasterRecoveryManager = null;
```
Comment (tkalkirill) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryController.java:49]: There will be no data races?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryFactory.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+import io.micronaut.context.annotation.Bean;
+import io.micronaut.context.annotation.Factory;
+import jakarta.inject.Singleton;
+import org.apache.ignite.internal.disaster.system.SystemDisasterRecoveryManager;
+import org.apache.ignite.internal.rest.RestFactory;
+
+/**
+ * Factory that defines beans required for the System Disaster Recovery Rest module.
+ */
```
Comment (tkalkirill) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryFactory.java:28]:
```suggestion
/** Factory that defines beans required for the System Disaster Recovery Rest module. */
```

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryFactory.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Bean
+    @Singleton
+    public SystemDisasterRecoveryManager systemDisasterRecoveryManager() {
+        return systemDisasterRecoveryManager;
+    }
+
+    @Override
+    public void cleanResources() {
+        systemDisasterRecoveryManager = null;
```
Comment (tkalkirill) [inline modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryFactory.java:45]: There will be no data races?

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java [hunk 7]
```diff
@@ -1159,6 +1160,7 @@ private RestComponent createRestComponent(String name) {
         Supplier<RestFactory> restManagerFactory = () -> new RestManagerFactory(restManager);
         Supplier<RestFactory> computeRestFactory = () -> new ComputeRestFactory(compute);
         Supplier<RestFactory> disasterRecoveryFactory = () -> new DisasterRecoveryFactory(disasterRecoveryManager);
+        Supplier<RestFactory> systemDisasterRecoveryFactory = () -> new SystemDisasterRecoveryFactory(systemDisasterRecoveryManager);
```
Comment (tkalkirill) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:1163]: How many suppliers are there...
Do you think it's possible to get rid of them somehow in another ticket?

File: modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryController.java (hunk 4 above)
Comment (rpuch) [reply modules/rest/src/main/java/org/apache/ignite/internal/rest/recovery/system/SystemDisasterRecoveryController.java:49]: This is called when the application is being stopped, there is a guarantee given by Micronaut that all its components are shut down. GC will see this assignment to null anyway.

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java (hunk 7 above)
Comment (rpuch) [reply modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteImpl.java:1163]: I think DI should do this. DI will come and save us :)

**Review by tkalkirill — APPROVED** [review]

**Pull request merged**